File: blind75/63_construct_binary_tree_from_preorder_and_inorder_traversal/construct_binary_tree_from_preorder_and_inorder_traversal.go
package construct_binary_tree_from_preorder_and_inorder_traversal

// TreeNode represents a node in a binary tree.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// BuildTree constructs a binary tree from preorder and inorder traversals.
// Uses recursive DFS with hashmap for O(n) time complexity.
func BuildTree(preorder []int, inorder []int) *TreeNode {
	// Create hashmap for O(1) lookup of inorder indices
	inorderMap := make(map[int]int)
	for i, val := range inorder {
		inorderMap[val] = i
	}

	preorderIndex := 0

	var build func(left, right int) *TreeNode
	build = func(left, right int) *TreeNode {
		// Base case: no elements to construct the tree
		if left > right {
			return nil
		}

		// Select the preorder index element as root and increment
		rootVal := preorder[preorderIndex]
		preorderIndex++

		root := &TreeNode{Val: rootVal}

		// Build left and right subtree
		// excluding inorderMap[rootVal] element because it's the root
		root.Left = build(left, inorderMap[rootVal]-1)
		root.Right = build(inorderMap[rootVal]+1, right)

		return root
	}

	return build(0, len(inorder)-1)
}

File: blind75/63_construct_binary_tree_from_preorder_and_inorder_traversal/construct_binary_tree_from_preorder_and_inorder_traversal.py
from typing import List, Optional


class TreeNode:
    """Definition for a binary tree node."""

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    """
    Constructs a binary tree from preorder and inorder traversals.
    Uses recursive DFS with hashmap for O(n) time complexity.

    Args:
        preorder: Preorder traversal of the tree
        inorder: Inorder traversal of the tree

    Returns:
        Root node of the constructed binary tree
    """
    # Create hashmap for O(1) lookup of inorder indices
    inorder_map = {val: i for i, val in enumerate(inorder)}
    preorder_index = 0

    def build(left: int, right: int) -> Optional[TreeNode]:
        nonlocal preorder_index

        # Base case: no elements to construct the tree
        if left > right:
            return None

        # Select the preorder index element as root and increment
        root_val = preorder[preorder_index]
        preorder_index += 1

        root = TreeNode(root_val)

        # Build left and right subtree
        # excluding inorder_map[root_val] element because it's the root
        root.left = build(left, inorder_map[root_val] - 1)
        root.right = build(inorder_map[root_val] + 1, right)

        return root

    return build(0, len(inorder) - 1)
